libp2p/muxers/mplex/lpchannel.nim

```diff
@@ -161,33 +172,49 @@ method closed*(s: LPChannel): bool =
   ## header of the file
   s.closedLocal
 
-method close*(s: LPChannel) {.async, gcsafe.} =
-  if s.closedLocal:
-    return
-
-  trace "closing local lpchannel", id = s.id,
-                                   initiator = s.initiator,
-                                   name = s.name,
-                                   oid = s.oid
-  # TODO: we should install a timer that on expire
-  # will make sure the channel did close by the remote
-  # so the hald-closed flow completed, if it didn't
-  # we should send a `reset` and move on.
-  await s.closeMessage()
-  s.closedLocal = true
-  if s.atEof: # already closed by remote close parent buffer imediately
-    await procCall BufferStream(s).close()
-
-  trace "lpchannel closed local", id = s.id,
-                                  initiator = s.initiator,
-                                  name = s.name,
-                                  oid = s.oid
-
-method reset*(s: LPChannel) {.base, async.} =
+method reset*(s: LPChannel) {.base, async, gcsafe.} =
   # we asyncCheck here because the other end
   # might be dead already - reset is always
   # optimistic
   asyncCheck s.resetMessage()
+  # # because of the async check above,
+  # # give the message time to depart
+  # await sleepAsync(100.millis)
   await procCall BufferStream(s).close()
   s.isEof = true
   s.closedLocal = true
+
+method close*(s: LPChannel) {.async, gcsafe.} =
+  if s.closedLocal:
+    trace "channel already closed", id = s.id,
+                                    initiator = s.initiator,
+                                    name = s.name,
+                                    oid = s.oid
+    return
+
+  proc closeRemote() {.async.} =
+    try:
+      trace "closing local lpchannel", id = s.id,
+                                       initiator = s.initiator,
+                                       name = s.name,
+                                       oid = s.oid
+      # TODO: we should install a timer that on expire
+      # will make sure the channel did close by the remote
+      # so the hald-closed flow completed, if it didn't
+      # we should send a `reset` and move on.
+      await s.closeMessage().wait(2.minutes)
+      s.closedLocal = true
+      if s.atEof: # already closed by remote close parent buffer imediately
```

Comment:
ah, this is the half-closed emulation logic, here is a break down of the flow:

- There are two methods, `close` and `closeRemote`
  - `close` is called when the channel is `conn.close()` locally
  - `closeRemote` is called when a close message is received from the remote
- Either of those can happen first - `close` or `closeRemote`
  - If the channel is closed locally, with `conn.close()`, then the `closedLocal` is set to `true` and you can't write to the channel anymore, but `EOF` is still `false` because the other end hasn't yet closed
  - if the channel is closed remotely with `closeRemote`, then the `EOF` is set to `true` and you can't read from the channel anymore, but `closeLocal` is still false, so you can write to it
- Now the third case is, what happen when we've been already `closeRemote`, and calling `conn.close()` locally? It means that we're completing the closing sequence and it's time to dispose of the channel, so that `EOF` flag is essentially checking if `closeRemote` was already called and we're completing the closing flow and should dispose of the channel.